Tab Bar and constraints (some screens)

MarcusNkadimeng opened the pull request:
Implemented a 3-tab tab bar, added a non-scrollable header to the transactions table, added SwiftLint to the project, added a file for UI specifications that I can use to declare custom components (e.g., colour) that I would like to use in the application. Modified the logic for authentication(signing up and logging in).

sjacobs1 reviewed (approved):
looks good to me

kieran-woodrow commented on budgetApp1/budgetApp1/AccountViewCell.swift:
```diff
@@ -18,8 +18,8 @@ class AccountViewCell: UITableViewCell {
         accountNameLabel.text = account.name
-        accountTypeLabel.text = account.type
-        balanceLabel.text = String(account.balance)
+        accountTypeLabel.text = "Cleared Balance (R)"
```
have a constants file for now that has all hardcoded text in it and reference it from there 😄

kieran-woodrow commented on budgetApp1/budgetApp1/AccountViewController.swift:
```diff
@@ -13,6 +13,8 @@ class AccountViewController: UIViewController {
     
     private lazy var viewModel = AccountViewModel(repository: AccountRepository(), delegate: self)
     
+    let uiSpecs = UISpecs()
```
what is uiSpecs? 😄 can it be private?

MarcusNkadimeng replied:
Done.

kieran-woodrow commented on budgetApp1/budgetApp1/AccountViewController.swift:
```diff
@@ -29,8 +31,10 @@ class AccountViewController: UIViewController {
+    
     func numberOfSections(in tableView: UITableView) -> Int {
-        return viewModel.AccountListCount
+        return viewModel.accountListCount
```
can omit return if 1 line. SwiftLint should be throwing a warning for that😄

kieran-woodrow commented on budgetApp1/budgetApp1/AccountViewController.swift:
```diff
@@ -55,8 +59,44 @@ extension AccountViewController: UITableViewDelegate, UITableViewDataSource {
         guard let cell = tableView.dequeueReusableCell(withIdentifier: AccountViewCell.identifier) as? AccountViewCell else { return UITableViewCell() }
         guard let account = viewModel.account(atIndex: indexPath.section) else { return UITableViewCell() }
         cell.populateWith(account: account)
+        cell.accountNameLabel.textColor = uiSpecs.appCustomColor2
```
this should be done in the cell itself. Call a function from the cell to do it.
Also if you are going to have a number in a name like this, appCustomColor2, spell the number out like Two😄

MarcusNkadimeng replied:
Done.

kieran-woodrow commented on budgetApp1/budgetApp1/AccountViewController.swift:
```diff
@@ -55,8 +59,44 @@ extension AccountViewController: UITableViewDelegate, UITableViewDataSource {
+        let cornerRadius: CGFloat = 18.0
+        cell.backgroundColor = UIColor.clear
+        
+        let layer = CAShapeLayer()
```
All of this logic should be handled by the cell, not by the view controller. create functions in the cell for it please😄

MarcusNkadimeng replied:
I changed it for the background colour but the other stuff like the cornerRadius have to be declared here

kieran-woodrow commented on budgetApp1/budgetApp1/AccountViewModel.swift:
```diff
@@ -16,31 +16,29 @@ class AccountViewModel {
-    var AccountListCount: Int {
-        return AccountList?.count ?? 0
+    var accountListCount: Int {
+        return accountList?.count ?? 0
```
can omit the return 😄

kieran-woodrow replied:
return statement still. Please read the warnings by swiftlint. If it is not showing a warning, something is wrong.
Apply else where

kieran-woodrow commented on budgetApp1/budgetApp1/AccountViewModel.swift:
```diff
@@ -16,31 +16,29 @@ class AccountViewModel {
     
     func account(atIndex: Int) -> Account? {
-        return AccountList?[atIndex] ?? nil
+        return accountList?[atIndex] ?? nil
```
same here. If swift lint is not picking it up, there is a problem with it. If it is, lease don't ignore the warnings😄

kieran-woodrow commented on budgetApp1/budgetApp1/LoginViewController.swift:
```diff
@@ -11,7 +11,8 @@ class LoginViewController: UIViewController {
     @IBOutlet weak var usernameField: UITextField!
```
private😄

kieran-woodrow commented on budgetApp1/budgetApp1/LoginViewController.swift:
```diff
@@ -11,7 +11,8 @@ class LoginViewController: UIViewController {
     @IBOutlet weak var passwordField: UITextField!
     
-    var user1 = user.init(username: "", password: "", fullname: "", emailAddress: "")
+    var user2 = User(username: "", password: "", fullname: "", emailAddress: "")
```
userTwo etc😄

kieran-woodrow commented on budgetApp1/budgetApp1/LoginViewController.swift:
```diff
@@ -11,7 +11,8 @@ class LoginViewController: UIViewController {
     
-    var user1 = user.init(username: "", password: "", fullname: "", emailAddress: "")
+    var user2 = User(username: "", password: "", fullname: "", emailAddress: "")
+    var user1 = User.init(username: "", password: "", fullname: "", emailAddress: "")
```
don't user User.init. Just use User() like you did above.😄

kieran-woodrow commented on budgetApp1/budgetApp1/LoginViewController.swift:
```diff
@@ -11,7 +11,8 @@ class LoginViewController: UIViewController {
+    var user2 = User(username: "", password: "", fullname: "", emailAddress: "")
+    var user1 = User.init(username: "", password: "", fullname: "", emailAddress: "")
     
     @IBAction func loginClicked(_ sender: Any) {
```
can this be private?

kieran-woodrow commented on budgetApp1/budgetApp1/LoginViewController.swift:
```diff
@@ -21,7 +22,6 @@ class LoginViewController: UIViewController {
         else {
-//            let alert2 = Alert.authenticationAlert(on: UIViewController)
             let alertController = UIAlertController(
                             title: "Invalid credentials",
```
Keep these strings in a constant file. Also, it's not really a swifty approach to throw an alert to say invalid credentials. Maybe use a more swift/ios approach 😄

MarcusNkadimeng replied:
Will do.

kieran-woodrow commented on budgetApp1/budgetApp1/Repository.swift:
```diff
@@ -26,18 +26,18 @@ protocol TransactionRepositoryType: AnyObject {
 class BudgetRepository: BudgetRepositoryType {
     
     func getBudgets(completion: @escaping (BudgetsResult)) {
```
please dont use the word get. use fetch. get is very java, fetch is more swift😄

MarcusNkadimeng replied:
Done

kieran-woodrow commented on budgetApp1/budgetApp1/Service Errors.swift:
```diff
@@ -63,3 +63,11 @@ class APIHandler {
+    static let budgetURL = baseURL + "budgets" + accessToken
+    static let accountsURL = baseURL + "budgets/2b0c25e3-0d5e-411d-a019-2522faa22c2b/accounts" + accessToken
+    static let transactionsURL = baseURL + "budgets/2b0c25e3-0d5e-411d-a019-2522faa22c2b/accounts/0515719a-f641-422b-8341-7073d5358dbe/transactions" + accessToken
+    
```
extra unnecessary new line 😄

kieran-woodrow commented on budgetApp1/budgetApp1/SignUpViewController.swift:
```diff
@@ -9,31 +9,16 @@ import UIKit
 
-    
     @IBOutlet weak var fullnameField: UITextField!
     @IBOutlet weak var emailAddressField: UITextField!
```
private 😄

kieran-woodrow commented on budgetApp1/budgetApp1/SignUpViewController.swift:
```diff
@@ -9,31 +9,16 @@ import UIKit
 
 class SignUpViewController: UIViewController {
 
-    
```
use marks. look at my other pr suggestions from kait or gia or seb

kieran-woodrow commented on budgetApp1/budgetApp1/SignUpViewController.swift:
```diff
@@ -9,31 +9,16 @@ import UIKit
     @IBOutlet weak var usernameField: UITextField!
     @IBOutlet weak var passwordField: UITextField!
     
     var signedUp = false
```
private

kieran-woodrow commented on budgetApp1/budgetApp1/SignUpViewController.swift:
```diff
@@ -9,31 +9,16 @@ import UIKit
-            print("Please sign up")
-        }
-    }
+    @IBAction func signedUpClicked(_ sender: Any) { }
```
empty function?

MarcusNkadimeng replied:
updated.

kieran-woodrow commented on budgetApp1/budgetApp1/SignUpViewController.swift:
```diff
@@ -43,12 +28,21 @@ class SignUpViewController: UIViewController {
+                loginViewController.user1 = primaryUser
+                signedUp = true
+            } else {
+                let alertController = UIAlertController(
```
same thing as before with alerts 😄

kieran-woodrow commented on budgetApp1/budgetApp1/TransactionTableViewCell.swift:
```diff
@@ -9,14 +9,14 @@ import UIKit
-    @IBOutlet weak var TransactionNameLabel: UILabel!
-    @IBOutlet weak var TransactionAmountLabel: UILabel!
+    @IBOutlet weak var transactionNameLabel: UILabel!
+    @IBOutlet weak var transactionAmountLabel: UILabel!
```
private

kieran-woodrow commented on budgetApp1/budgetApp1/TransactionTableViewCell.swift:
```diff
@@ -9,14 +9,14 @@ import UIKit
     
     static let identifier = "TransactionTableViewCell"
     
-    func populateWith(transaction: Transaction){
```
static let should be in a constants file

kieran-woodrow commented on budgetApp1/budgetApp1/TransactionViewController.swift:
```diff
@@ -9,6 +9,7 @@ import UIKit
 
+    let uiSpecs = UISpecs()
     
     @IBOutlet weak var tableView: UITableView!
```
private

kieran-woodrow commented on budgetApp1/budgetApp1/TransactionViewController.swift:
```diff
@@ -9,6 +9,7 @@ import UIKit
 
 class TransactionViewController: UIViewController {
 
+    let uiSpecs = UISpecs()
```
why is this being used again?

MarcusNkadimeng replied:
for the table border colour. Instead of always typing out that long custom colour I just call it from the uiSpecs class I created

kieran-woodrow commented on budgetApp1/budgetApp1/TransactionViewController.swift:
```diff
@@ -18,21 +19,23 @@ class TransactionViewController: UIViewController {
     
     private func setupTableView() {
         tableView.delegate = self
         tableView.dataSource = self
```
can be in a function called setUptableView

kieran-woodrow commented on budgetApp1/budgetApp1/TransactionViewController.swift:
```diff
@@ -18,21 +19,23 @@ class TransactionViewController: UIViewController {
 extension TransactionViewController: UITableViewDelegate, UITableViewDataSource {
     func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return viewModel.TransactionListCount
+        return viewModel.transactionListCount
```
omit reutn

MarcusNkadimeng replied:
Done.

kieran-woodrow commented on budgetApp1/budgetApp1/UISpecs.swift:
```diff
@@ -0,0 +1,15 @@
+
+class UISpecs {
+    let appCustomColor = UIColor(red: 130/255.0, green: 120/255.0, blue: 225/255.0, alpha: 1.0)
+    let appCustomColor2 = UIColor(red: 61/255.0, green: 28/255.0, blue:230/255.0, alpha: 1.0)
```
appCustomColoTwo
rather call it primary and secondary colours

MarcusNkadimeng replied:
Good suggestion!

kieran-woodrow commented:
> looks good to me

what bout 19 files changed? Keep each other accountable. The rule is between 5 and 10.. Please keep to this standard. In future PR's it will just get rejected straight away because the review process is too long if there are a lot of files😄

MarcusNkadimeng commented:
Added changes received via feedback.
• Created a file for identifiers and grouped them in the corresponding structs. 
• Privatised variables that needed to be privatised.
• Updated variable names where needed.

kieran-woodrow commented on budgetApp1/budgetApp1/AccountViewCell.swift:
```diff
@@ -9,21 +9,21 @@ import UIKit
     
     static func nib() -> UINib {
-        return UINib(nibName: "AccountViewCell", bundle: nil)
+        UINib(nibName: "AccountViewCell", bundle: nil)
```
This cell name should be in the constants file

kieran-woodrow replied:
see this was resolved but wasn't changed?

kieran-woodrow commented on budgetApp1/budgetApp1/AccountViewController.swift:
```diff
@@ -52,11 +54,45 @@ extension AccountViewController: UITableViewDelegate, UITableViewDataSource {
+    
+    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+        let cornerRadius: CGFloat = 18.0
+        
```
unnecessary new line

kieran-woodrow commented on budgetApp1/budgetApp1/AccountViewController.swift:
```diff
@@ -52,11 +54,45 @@ extension AccountViewController: UITableViewDelegate, UITableViewDataSource {
+            pathRef.addRoundedRect(in: bounds, cornerWidth: cornerRadius, cornerHeight: cornerRadius)
+        } else if indexPath.row == 0 {
+            pathRef.move(to: CGPoint(x: bounds.minX, y: bounds.maxY))
+            pathRef.addArc(tangent1End: CGPoint(x: bounds.minX, y: bounds.minY), tangent2End: CGPoint(x: bounds.midX, y: bounds.midY), radius: cornerRadius)
```
all of this logic should be in the cell

MarcusNkadimeng replied:
May I please change this later? I'be been getting a lot of errors trying to fix this? Can I do it in the next PR?

rvanderwesthuizen replied:
@MarcusNkadimeng @kieran-woodrow 
This is Kieran's comment so you need to discuss with him, but if you do leave this for a next PR please add a warning above
For example
#warning("Move this to cell")

kieran-woodrow replied:
happy for you to change it in the next PR so we can get this one through. But please change it

MarcusNkadimeng replied:
Okay

kieran-woodrow commented on budgetApp1/budgetApp1/BudgetViewController.swift:
```diff
@@ -23,9 +23,7 @@ class BudgetViewController: UIViewController {
         super.viewDidLoad()
         
         viewModel = BudgetViewModel(repository: BudgetRepository(), delegate: self)
```
The didset on the viewModel is a bit strange. Can you not instantiate the view model normally and then call the fillUI in the viewDidLoad?

kieran-woodrow commented on budgetApp1/budgetApp1/Constants.swift:
```diff
@@ -0,0 +1,28 @@
+
+// MARK: - nib identifiers
+struct NibIdentifiers {
+    static let AccountViewCellIdentifier = "AccountViewCell"
```
these variables can still follow camel case. 
accountViewCellIdentifier, transactionViewCellIdentifier

kieran-woodrow replied:
this was resolved but not changed?
Please do not mark the conversation s resolved until the change is made or you provide feedback as to why you are going to keep it?

kieran-woodrow replied:
still resolved but not chaged or no comment..?

MarcusNkadimeng replied:
Done.

kieran-woodrow commented on budgetApp1/budgetApp1/LoginViewController.swift:
```diff
@@ -8,20 +8,19 @@
+    @IBOutlet private weak var passwordField: UITextField!
     
-    var user1 = user.init(username: "", password: "", fullname: "", emailAddress: "")
+    var userTwo = User(username: "", password: "", fullname: "", emailAddress: "")
```
good jib renaming the 2 to Two, but why is there even a user Two? why not just user?

kieran-woodrow replied:
and this type of logic of instattiating a User should be done in the viewModel for this screen

MarcusNkadimeng replied:
I think I'll change it later with storing user data using coredata.

kieran-woodrow commented on budgetApp1/budgetApp1/LoginViewController.swift:
```diff
@@ -8,20 +8,19 @@
         guard let password = passwordField.text else { return }
         
-        if (username == user1.username && password == user1.password) {
+        if (username == userTwo.username && password == userTwo.password) {
```
should be calling a function from the viewModel. This is logic. Logic should not be done in the viewController

kieran-woodrow commented on budgetApp1/budgetApp1/Repository.swift:
```diff
@@ -26,18 +26,18 @@ protocol TransactionRepositoryType: AnyObject {
 class BudgetRepository: BudgetRepositoryType {
     
```
get functions hould be fetch. Please keep in mind oging forward

kieran-woodrow replied:
after this change

kieran-woodrow commented on budgetApp1/budgetApp1/SignUpViewController.swift:
```diff
@@ -9,46 +9,45 @@ import UIKit
-    @IBOutlet weak var emailAddressField: UITextField!
-    @IBOutlet weak var usernameField: UITextField!
-    @IBOutlet weak var passwordField: UITextField!
+    // MARK: - signUp token
```
mark can be called variables to indicate variables below. Not a specific variable. Please apply marks to all other places in these files that you have touched in this PR

kieran-woodrow commented on budgetApp1/budgetApp1/SignUpViewController.swift:
```diff
@@ -9,46 +9,45 @@ import UIKit
     
+    // MARK: - Segue to validate credentials and log in
     override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
         if segue.identifier == "signUpSegue" {
```
sugue name in constants file please

kieran-woodrow commented on budgetApp1/budgetApp1/SignUpViewController.swift:
```diff
@@ -9,46 +9,45 @@ import UIKit
+    // MARK: - Segue to validate credentials and log in
     override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
         if segue.identifier == "signUpSegue" {
             guard let username = usernameField.text else { return }
```
this is logic. Please do this in the viewController

kieran-woodrow replied:
also signUpSegue should be in constants file

kieran-woodrow commented on budgetApp1/budgetApp1/SignUpViewController.swift:
```diff
@@ -9,46 +9,45 @@ import UIKit
+                loginViewController.userTwo = primaryUser
+            } else {
+                let alertController = UIAlertController(
+                                title: "Incomplete user details",
```
strings in constant file please. Please keep this in mind so i don't always retype it. All strings should be in. strings or constant file

kieran-woodrow commented on budgetApp1/budgetApp1/SignUpViewController.swift:
```diff
@@ -9,46 +9,45 @@ import UIKit
+                                preferredStyle: .alert
+                            )
+                alertController.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: "Default action"), style: .default, handler: { _ in
+                    NSLog("The \"OK\" alert occured.")
```
why are you using NSLog? This is logging to the console. Please do propper error management..

kieran-woodrow commented on budgetApp1/budgetApp1/TransactionViewModel.swift:
```diff
@@ -15,33 +15,31 @@ protocol TransactionViewModelDelegate: AnyObject {
-    var TransactionListCount: Int {
-        return TransactionList?.count ?? 0
+    var transactionListCount: Int {
+        return transactionList?.count ?? 0
```
returns..please apply these generic comments everywhere else

kieran-woodrow commented on budgetApp1/budgetApp1/UISpecs.swift:
```diff
@@ -0,0 +1,15 @@
+//  Created by Lehlohonolo Nkadimeng on 2024/04/20.
+//
+
+import Foundation
```
Do you need both foundations and UIKit? I think you can remove foundations

simekanidev commented:
Please use PR template. Also please commit more frequently.

kieran-woodrow commented on budgetApp1/budgetApp1/BudgetViewController.swift:
```diff
@@ -23,9 +23,7 @@ class BudgetViewController: UIViewController {
+        viewModel?.fetchBudgets()
     }
 
     private func fillUI() {
```
this name is a bit misleading. Try Avoid UI. Apple uses UI for UIKit classes. so fillUI made me immediately think it was something to do with a UI class.  
Perhaps use setUpView. 
and this is still logic. Create functions in the viewModel please.
and please make the IBOutlets private

MarcusNkadimeng replied:
Done

MarcusNkadimeng commented:
## Description 
Logic that was in the Login, SignUp, and Account ViewControllers has been moved to the respective ViewModels.
Added MARKs to the relevant code sections.
Created a UserViewModel to handle signup and login logic.
Created a file of strings for nib identifiers and any other stuff that is a string in the project.
Used better naming conventions.|
Functions inherited from the viewModelDelegates still need implementation. To be done at a later stage.
 
## Screenshots/GIFs 
 
## Checklist 
Please review and check the following before submitting this pull request: 
- [x] The code follows the project's coding standards. 
- [ ] 100% Unit test coverage for View Models 
- [ ] All warnings Resolved.

kieran-woodrow commented on budgetApp1/budgetApp1/BudgetViewController.swift:
```diff
@@ -9,49 +9,41 @@ import UIKit
     private var viewModel: BudgetViewModel? {
         didSet {
-            fillUI()
+            viewModel?.fetchBudgets()
```
This is still a bit strange.
 viewModel?.fetchBudgets() is something that typically belongs in the viewDidLoad()
 And then you can instantiate this as private var viewModel = BudgetViewModel()

MarcusNkadimeng replied:
Changed it.

kieran-woodrow replied:
resolved

kieran-woodrow commented on budgetApp1/budgetApp1/LoginViewController.swift:
```diff
@@ -1,45 +1,36 @@
+    @IBOutlet private weak var usernameField: UITextField!
+    @IBOutlet private weak var passwordField: UITextField!
+    
+    let userVM = UserViewModel.shared // Use shared instance
```
please don't abbreviate. 
let userViewModel

kieran-woodrow replied:
also, this is very bad. Please don't use a singleton. 
UserViewModel.shared is a singleton. You shouldn't have to do this.

MarcusNkadimeng replied:
I am using a singleton because for some reason when I call the user stored in the userVM in the Login screen, it is empty. I think it is because when I instantiate the userVM again in the login after the user has signed in, it loses the value stored in User so I thought using a singleton would make the details persistent across both Login and SignUp Controllers. 
I know it's bad coding principles but I just had to find a way around this.

kieran-woodrow replied:
lets have a call. I'm sure we can get it working the right way :)

MarcusNkadimeng replied:
Okay

kieran-woodrow replied:
had a chat with you on a call. We will use coredata in the next PR

kieran-woodrow commented on budgetApp1/budgetApp1/LoginViewController.swift:
```diff
@@ -1,45 +1,36 @@
-
     }
+    
     override func viewDidLoad() {
```
can remove this viewDidLoad if you are not using it

MarcusNkadimeng replied:
Done. Will do this for all ViewControllers.

kieran-woodrow commented on budgetApp1/budgetApp1/SignUpViewController.swift:
```diff
@@ -1,56 +1,27 @@
 
+    // MARK: - IBOutlets
+    @IBOutlet private weak var fullnameField: UITextField!
+    @IBOutlet private weak var emailAddressField: UITextField!
```
you can remove the word Field in these vars

MarcusNkadimeng replied:
DONE

kieran-woodrow commented on budgetApp1/budgetApp1/SignUpViewController.swift:
```diff
@@ -1,56 +1,27 @@
-    
-    var signedUp = false
+    // MARK: - Variables
+    let userVM = UserViewModel.shared // Use shared instance
```
same comment as before. Please address :)

kieran-woodrow replied:
had a chat with you on a call. We will use coredata in the next PR

kieran-woodrow commented on budgetApp1/budgetApp1/UserViewModel.swift:
```diff
@@ -0,0 +1,21 @@
+class UserViewModel {
+    
+    // MARK: - Singleton instantiation of ViewModel
+    static let shared = UserViewModel()
```
avoid this :) Singletons should be avoided

MarcusNkadimeng replied:
I am using a singleton because for some reason when I call the user stored in the userVM in the Login screen, it is empty. I think it is because when I instantiate the userVM again in the login after the user has signed in, it loses the value stored in User so I thought using a singleton would make the details persistent across both Login and SignUp Controllers.
I know it's bad coding principles but I just had to find a way around this.

kieran-woodrow replied:
had a chat with you on a call. We will use coredata in the next PR

rvanderwesthuizen commented on budgetApp1/budgetApp1/Constants.swift:
```diff
@@ -0,0 +1,40 @@
+
+// MARK: - UI Stuff
+struct UIComponents {
+    static let accountTypeLabel = "Cleared Balance (R)"
```
@MarcusNkadimeng 
Please consider renaming accountTypeLabel since this is now used for cleared balance and not the account type

MarcusNkadimeng replied:
done

rvanderwesthuizen commented on budgetApp1/budgetApp1/AccountViewCell.swift:
```diff
@@ -9,21 +9,24 @@ import UIKit
         accountNameLabel.text = account.name
-        accountTypeLabel.text = account.type
-        balanceLabel.text = String(account.balance)
+        accountTypeLabel.text = UIComponents.accountTypeLabel
```
@MarcusNkadimeng
Please consider renaming accountTypeLabel since this is now used for cleared balance and not the account type

MarcusNkadimeng replied:
Done

rvanderwesthuizen commented on budgetApp1/budgetApp1/AccountViewCell.swift:
```diff
@@ -9,21 +9,24 @@ import UIKit
     
+    // MARK: - IBOutlets
+    @IBOutlet private weak var accountNameLabel: UILabel!
+    @IBOutlet private weak var accountTypeLabel: UILabel!
```
@MarcusNkadimeng
Please consider renaming accountTypeLabel since this is now used for cleared balance and not the account type

MarcusNkadimeng replied:
Done

rvanderwesthuizen replied:
Do you still need to push this change?
Please don't mark things as resolved unless you've pushed the changes so we can see them

MarcusNkadimeng replied:
I haven't pushed it but I have changed it. I'll push it after my call with Kieran regarding the singleton used in SignUpViewController

rvanderwesthuizen commented on budgetApp1/budgetApp1/BudgetViewController.swift:
```diff
@@ -9,49 +9,41 @@ import UIKit
     
     func reloadView() {
-        fillUI()
+        guard let budget = viewModel?.budgetList?.first else {
+            return
+        }
```
Please move the return into the same line
Due to the small else block this is a bit cleaner and saves lines

```suggestion
        guard let budget = viewModel?.budgetList?.first else { return }
```

MarcusNkadimeng replied:
Done.

MarcusNkadimeng commented:
## Description 
Improved logic in the budgetViewController.
Renamed input fields in login and signup screens.
 
## Screenshots/GIFs 
 
## Checklist 
Please review and check the following before submitting this pull request: 
- [x] The code follows the project's coding standards. 
- [ ] 100% Unit test coverage for View Models 
- [ ] All warnings Resolved.

rvanderwesthuizen commented on budgetApp1/budgetApp1/BudgetViewController.swift:
```diff
@@ -9,49 +9,35 @@ import UIKit
-    @IBOutlet weak var budgetNameLabel: UILabel!
-    @IBOutlet weak var endDateLabel: UILabel!
-    @IBOutlet weak var startDateLabel: UILabel!
+    private var viewModel: BudgetViewModel?
```
Please declare your viewmodel as you did on TransactionViewController

With the instantiation and lazy included

MarcusNkadimeng replied:
Done.

rvanderwesthuizen commented on budgetApp1/budgetApp1/BudgetViewController.swift:
```diff
@@ -9,49 +9,35 @@ import UIKit
     override func viewDidLoad() {
         super.viewDidLoad()
-        
         viewModel = BudgetViewModel(repository: BudgetRepository(), delegate: self)
```
Then along with the above you remove this

MarcusNkadimeng replied:
Done.
